Inject a default organization into groups objects without organizations

lyzadanger opened the pull request:
The way the client app is currently structured, things fall apart if a given group doesn't have an associated organization object with a logo. I spent quite some time and some dead-end branches trying to be more clever about a fallback logo and situation when a group doesn't have an organization, but it got wildly complex, almost immediately.

This approach felt like the safest way to make the client tolerate groups with missing organization relations until we can look at the bigger picture here. Any group without an organization will be treated as if it is within the default, Hypothesis organization—and will carry the Hypothesis logo, for now.

This will allow us to support third-party private groups in the client in the short term—and we're under some time pressure here.

This PR does _not_ have any dependencies on changes in `h` (though the inverse is most definitely not true).

seanh commented:
> Any group without an organization will be treated as if it is within the default, Hypothesis organization—and will carry the Hypothesis logo, for now.

FWIW I don't think there's any "for now" about showing them with the Hypothesis name and logo. Groups appearing under the Hypothesis logo and name is as it was designed. Whether the implementation of this involves a "Hypothesis organization" in the model or not is a different question though. But the "mental model" that the UI designs were intended to communicate to the user was that a group either "belongs to" a BioPub or eLife etc or, by default, a group "belongs to" Hypothesis:

![39471275-402ff3ee-4d3a-11e8-9f83-3fa4fbf3253e](https://user-images.githubusercontent.com/22498/44204107-49582c80-a149-11e8-87b9-fe2cca4df603.png)

![37093505-7131afcc-2207-11e8-9ca2-f28f5512db08](https://user-images.githubusercontent.com/22498/44204137-67259180-a149-11e8-92d3-568246684fb3.png)

<https://github.com/hypothesis/product-backlog/issues/502>, <https://github.com/hypothesis/product-backlog/issues/497>

dwhly commented:
Interesting discussion and good grist for a new designer. The kind of nesting (shown above in the mocks that sean referenced) might just as easily be thought of as grouping things which are under a service (or in this case under a separate namespace/3rd party auth regime), vs under an organization per-se.

How (or whether) we graphically represent that distinction differently is something to reflect on.

lyzadanger commented:
I've spent some time today testing this branch against several branches on `h` and in both first-party and third-party contexts. Here are some test results for third-party context, specifically the "publisher test site" locally.

I'm primarily seeding data using https://github.com/hypothesis/devdata, but extending it with some additional groups and metadata in some cases.

### Third-party site, `organization` still required

Running this branch against `h` `master`—that is, at time of writing, with `organization` yet a required relation on groups results in this:

<img width="478" alt="screen shot 2018-08-16 at 11 37 47 am" src="https://user-images.githubusercontent.com/439947/44232160-16b43100-a155-11e8-9cf6-8d06c838ba98.png">

(Note: I added a logo to the Partner.org organization for clarity here). There is a "Partner Public" group seeded from devdata, and I added a restricted group as well to show how these would show up.

There are no private groups visible because `master` on `h` does not yet have support for third-party private groups...

Thus, I believe this experience is what we'd expect...

### Third-party site, against an `h` branch that includes commits from both https://github.com/hypothesis/h/pull/5223 and https://github.com/hypothesis/h/pull/5218

This setup represents "where we're trying to get to ultimately", merging in support for optional organizations and the ability to add a third-party group via the API.

With this configuration, I'm able to create some third-party private groups via the API (I use Postman for this) "on behalf of" a user within the "partner.org" authority. Logged in as that user, I now see:

<img width="499" alt="screen shot 2018-08-16 at 11 26 28 am" src="https://user-images.githubusercontent.com/439947/44232377-ba054600-a155-11e8-85a3-aea9a794cea4.png">

Where the last two groups are private groups in the third-party authority. These private groups have no organization relationship.

This, too, is as I'd expect it.

### Third-party site, against an `h` branch that removes the organization requirement, but excludes https://github.com/hypothesis/h/pull/5218

i.e. the site will allow groups with no organization, but there's no mechanism to create said groups easily as that is part of https://github.com/hypothesis/h/pull/5218.

In any case, this setup works fine, too:

<img width="451" alt="screen shot 2018-08-16 at 11 30 05 am" src="https://user-images.githubusercontent.com/439947/44232526-3435ca80-a156-11e8-86fa-0f4c782f50bc.png">

I'll cover first-party group testing in another comment as this is getting crazy long.

lyzadanger commented:
And first-party context, tested locally...

### On `h` master, `organization` required still...

When I add a private group at this stage...

<img width="341" alt="screen shot 2018-08-16 at 1 17 18 pm" src="https://user-images.githubusercontent.com/439947/44232897-482dfc00-a157-11e8-9055-77bb53e2bb24.png">

`h` will automatically associate my new group "With Org" with the "default" organization (Hypothesis): 

<img width="480" alt="screen shot 2018-08-16 at 1 18 23 pm" src="https://user-images.githubusercontent.com/439947/44232927-57ad4500-a157-11e8-9e9c-c76c39e19d45.png">

All good...

### Against `h` with commits from https://github.com/hypothesis/h/pull/5218 and https://github.com/hypothesis/h/pull/5223

So now we're able to create groups without an organization...if I create a private group _now_, `h` will not associate it with an organization. That means we end up with a mixed state in the database—my earlier-created "With Org" group is attached to the default organization, while "Without Org" and "Also without Org" have no organization.

From the UI's perspective, though, they look cohesively fine—I added "Also without Org" to show that the client correctly alphabetizes and presents groups that are in the default org along with groups that have no org, for now:

<img width="463" alt="screen shot 2018-08-16 at 1 20 16 pm" src="https://user-images.githubusercontent.com/439947/44233061-af4bb080-a157-11e8-8995-dec7cc6ce972.png">

We'll be cleaning up the database soon after we deploy some of these changes to make the organization relationships consistent, but during the short term, there will be some mixed records—some with and some without organizations—so I've gone to fair trouble to make sure different mixes won't bother the client.

robertknight commented on src/sidebar/services/test/groups-test.js:
```diff
@@ -205,6 +205,21 @@ describe('groups', function() {
+      var svc = service();
+      const groups = [
+        { id: '39r39f', name: 'Ding Dong!' },
+      ];
```
Our client code uses a mixture of `var` (older files) and `const`/`let` (newer files). I've been following the "be consistent with surrounding code" policy, but this is still a cause for confusion. There are tools which can "modernize" JS codebases by applying transformations including `var` => `const`/`let` which would avoid the confusion. All I need is a 👍 from other devs that are happy for us to make this kind of sweeping change.
